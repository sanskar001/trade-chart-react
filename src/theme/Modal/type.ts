type CloseHandlerType = () => void;

export interface BackDropProps {
  backdropClass?: string;
  onClose: CloseHandlerType;
}

export interface ModalWindowProps {
  children: React.ReactNode;
  onClose: CloseHandlerType;
  headerText?: string;
  modalClass?: string;
  footerText?: string;
}

export interface ModalProps extends BackDropProps, ModalWindowProps {
  isShowModal: boolean;
}
