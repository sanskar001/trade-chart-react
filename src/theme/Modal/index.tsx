import React from "react";
import { createPortal } from "react-dom";
import Button from "../Button";
import CloseICon from "../../components/SVG/CloseIcon";
import { formatClass } from "../../util/formatClass";
import { BackDropProps, ModalProps, ModalWindowProps } from "./type";

const BackDrop: React.FC<BackDropProps> = ({ backdropClass = "", onClose }) => (
  <div
    className={formatClass(`backdrop ${backdropClass}`)}
    onClick={onClose}
  ></div>
);

const ModalWindow: React.FC<ModalWindowProps> = ({
  children,
  onClose,
  headerText = "",
  modalClass = "",
  footerText = "",
}) => (
  <div className={formatClass(`modal ${modalClass}`)}>
    {headerText && (
      <div className="modal-header">
        <h4>{headerText}</h4>
        <Button className="p-1" onClick={onClose}>
          <CloseICon />
        </Button>
      </div>
    )}
    <div className="modal-body">{children}</div>
    {footerText && <div className="modal-footer">{footerText}</div>}
  </div>
);

const Modal: React.FC<ModalProps> = ({
  children,
  isShowModal = false,
  onClose,
  headerText = "",
  modalClass = "",
  backdropClass = "",
  footerText = "",
}) => {
  const chartWidgetContainer = document.getElementById("chart-widget");

  return (
    <React.Fragment>
      {isShowModal &&
        createPortal(
          <ModalWindow
            onClose={onClose}
            modalClass={modalClass}
            headerText={headerText}
            footerText={footerText}
          >
            {children}
          </ModalWindow>,
          chartWidgetContainer as Element
        )}
      {isShowModal &&
        createPortal(
          <BackDrop onClose={onClose} backdropClass={backdropClass} />,
          chartWidgetContainer as Element
        )}
    </React.Fragment>
  );
};

export default Modal;
