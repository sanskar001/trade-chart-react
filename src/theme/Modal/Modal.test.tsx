import React from "react";
import renderer from "react-test-renderer";
import Modal from ".";

const closeHandler = jest.fn();

jest.mock("react-dom", () => {
  const original = jest.requireActual("react-dom");
  return {
    ...original,
    createPortal: (node: React.ReactNode) => node,
  };
});

it("renders correctly when there is default modal", () => {
  const tree = renderer
    .create(
      <Modal isShowModal={true} onClose={closeHandler}>
        <div>
          <h1>Modal</h1>
          <p>
            Lorem ipsum, dolor sit amet consectetur adipisicing elit. Molestias,
            labore.
          </p>
        </div>
      </Modal>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly when there is modal custom class", () => {
  const tree = renderer
    .create(
      <Modal
        isShowModal={true}
        onClose={closeHandler}
        modalClass="bg-green-100"
        backdropClass="bg-blue-400/20"
      >
        <div>
          <h1>Modal</h1>
          <p>
            Lorem ipsum, dolor sit amet consectetur adipisicing elit. Molestias,
            labore.
          </p>
        </div>
      </Modal>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly when there is modal with header and footer", () => {
  const tree = renderer
    .create(
      <Modal
        isShowModal={true}
        onClose={closeHandler}
        headerText="Modal Header"
        footerText="Modal Footer"
      >
        <div>
          <h1>Modal</h1>
          <p>
            Lorem ipsum, dolor sit amet consectetur adipisicing elit. Molestias,
            labore.
          </p>
        </div>
      </Modal>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
