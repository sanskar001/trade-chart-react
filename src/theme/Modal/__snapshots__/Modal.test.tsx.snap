// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there is default modal 1`] = `
Array [
  <div
    className="modal"
  >
    <div
      className="modal-body"
    >
      <div>
        <h1>
          Modal
        </h1>
        <p>
          Lorem ipsum, dolor sit amet consectetur adipisicing elit. Molestias, labore.
        </p>
      </div>
    </div>
  </div>,
  <div
    className="backdrop"
    onClick={[MockFunction]}
  />,
]
`;

exports[`renders correctly when there is modal custom class 1`] = `
Array [
  <div
    className="modal bg-green-100"
  >
    <div
      className="modal-body"
    >
      <div>
        <h1>
          Modal
        </h1>
        <p>
          Lorem ipsum, dolor sit amet consectetur adipisicing elit. Molestias, labore.
        </p>
      </div>
    </div>
  </div>,
  <div
    className="backdrop bg-blue-400/20"
    onClick={[MockFunction]}
  />,
]
`;

exports[`renders correctly when there is modal with header and footer 1`] = `
Array [
  <div
    className="modal"
  >
    <div
      className="modal-header"
    >
      <h4>
        Modal Header
      </h4>
      <button
        className="btn p-1"
        disabled={false}
        onClick={[MockFunction]}
        type="button"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.017 4l-7.034 7.025-7-6.992L4 5.017 11 12l-7 6.992.983.975 7-6.984L19.017 20l.983-.975L12.967 12 20 4.983 19.017 4z"
            fill="#131722"
          />
        </svg>
      </button>
    </div>
    <div
      className="modal-body"
    >
      <div>
        <h1>
          Modal
        </h1>
        <p>
          Lorem ipsum, dolor sit amet consectetur adipisicing elit. Molestias, labore.
        </p>
      </div>
    </div>
    <div
      className="modal-footer"
    >
      Modal Footer
    </div>
  </div>,
  <div
    className="backdrop"
    onClick={[MockFunction]}
  />,
]
`;
